login.js

```diff
@@ -1,31 +1,42 @@
-let isAMatch;
-
-function verifiyPassword(hash) {
-  const passwords = ['greenbear757', 'organicswan134', 'bluemouse519', 'heavytiger886', 'heavyladybug575'];
-
-  isAMatch = [];
-
-  passwords.forEach(
-    (item) => checkpw(item, hash,
-    (result) => {return isAMatch.push(result);}, null));
-}
-
-function displayResult() {
-  const output = document.getElementById('validation');
-
-  return isAMatch.some(
-           (match) => match === true) ?
-            output.textContent = 'Welcome Back!' :
-            output.textContent = 'Invalid password';
-}
-
-
-document.getElementById('submit').addEventListener('click', function() {
-
-  let userPassword = document.getElementById('password_field').value;
-  let salt = gensalt(5);
-
-  hashpw(userPassword, salt, verifiyPassword, null);
-  setTimeout(displayResult, 500);
-
-});
+(function () {
+  let passwordChecker = {
+    _passwords: ['greenbear757', 'organicswan134', 'bluemouse519', 'heavytiger886', 'heavyladybug575'],
+    _isAMatch: [],
+
+    get isAMatch() {
+      return this._isAMatch;
+    },
+
+    verifiyPassword(hash) {
+      passwordChecker._isAMatch = [];
+      let storeResults = (result) => {return passwordChecker._isAMatch.push(result);};
+      passwordChecker._passwords.forEach((item) => checkpw(item, hash,storeResults , null));
+    }
+  };
+
+  const ui = {
+    _userPassword: document.getElementById('password_field'),
+    _output: document.getElementById('validation'),
+    _salt: gensalt(5),
+
+    get userPassword() {
+      return this._userPassword.value;
+    },
+
+    displayResult() {
+      return passwordChecker.isAMatch.some((match) => match) ?
+      ui._output.textContent = 'Welcome Back!' :
+      ui._output.textContent = 'Invalid password';
+    },
+
+    clickStart() {
+      document.getElementById('submit').addEventListener('click', () => {
+        hashpw(this.userPassword, this._salt, passwordChecker.verifiyPassword, null);
```

Comment:
`hashpw` and `gensalt` seem to be knowledge that belongs to the `passwordChecker` rather than the `ui`.
If the encryption library gets an update and suddenly you have to change the way you use `hashpw`, you will have to go to the user interface to make changes, which is not intuitive, as the ui should only deal with user interface stuff.

At the moment you have `hashpw` in the `ui` and `checkpw` in `passwordChecker`, so the knowledge of the library functions is spread all over the application. It's better to encapsulate this knowledge in just one object, and have that object deal with the library. If something changes in the library, only one object needs to be updated! :slightly_smiling_face: 

This is related with the [Single Responsibility Principle](https://levelup.gitconnected.com/heres-a-list-of-5-coding-principles-that-help-you-write-better-code-d9b8ceb63c1b): your objects should have a well defined role, a **single responsibility**, a single reason to change.

So it is better to delegate this behaviour to the `passwordChecker`, in a method that calculates the salt and calls the library function. Then the `ui` can ask the `passwordChecker` "_hey, can you do this thing for me, you know how to do it, but I don't_". It would ask it like this: `passwordChecker.METHOD_TO_CHECK_PASSWORD(this.userPassword)`.